### PR TITLE
[wrangler] teach wrangler to search docs

### PR DESCRIPTION
**Pull request by rozenmd**
I basically taught wrangler how to do a "I'm feeling lucky" search

How it works:
- The wrangler user enters `wrangler docs <command>`
- wrangler sends an API request to Algolia's index of https://developers.cloudflare.com/workers/wrangler/
- wrangler sends the user to the first URL in the index that matches the search term

Here's what it looks like in action:

https://user-images.githubusercontent.com/3822106/207370613-eacccce9-ddea-4c9b-82d1-de476ff52fa2.mov

@rita3ko FYI

### CODE REVIEW

File: packages/wrangler/src/docs/constants.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const ALGOLIA_APPLICATION_ID = "8MU1G3QO9P";
+export const ALGOLIA_API_KEY = "92ece5213bea0489b4a5a4c21c7e916c";
```
Comment (rozenmd) [inline packages/wrangler/src/docs/constants.ts:2]: These are public keys FYI

See also: https://github.com/cloudflare/cloudflare-docs/blob/b835012e7160777fe3d611e585d729860514139a/assets/search.ts

File: packages/wrangler/src/docs/constants.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export const ALGOLIA_APPLICATION_ID = "8MU1G3QO9P";
+export const ALGOLIA_API_KEY = "045e8dbec8c137a52f0f56e196d7abe0";
```
Comment (rozenmd) [inline packages/wrangler/src/docs/constants.ts:2]: These are public keys FYI

See also: https://github.com/cloudflare/cloudflare-docs/blob/b835012e7160777fe3d611e585d729860514139a/assets/search.ts

**Review by JacobMGEvans — APPROVED** [review]
I am all for this once the security thing is cleared up or approved.

File: packages/wrangler/src/docs/index.ts [hunk 3]
```diff
@@ -1,89 +1,49 @@
...
+				}),
+				headers: {
+					"X-Algolia-API-Key": key,
+					"X-Algolia-Application-Id": id,
+				},
+			}
+		);
+		const searchData = (await searchResp.json()) as { hits: { url: string }[] };
+		logger.debug("searchData: ", searchData);
+		if (searchData.hits[0]) {
```
Comment (rozenmd) [inline packages/wrangler/src/docs/index.ts:44]: I wonder what happens when there are no hits...

File: packages/wrangler/src/docs/index.ts [hunk 4]
```diff
@@ -1,89 +1,53 @@
...
-		urlToOpen = "https://developers.cloudflare.com/workers/wrangler/api/";
-	} else if (args.command && isValidParam(args.command)) {
-		//otherwise, they get the wrangler commands page
-		urlToOpen += `#${argToUrlHash[args.command]}`;
+	let urlToOpen = "";
+
+	if (args.command) {
+		const id = "8MU1G3QO9P";
+		const index = "developers-cloudflare2";
+		const key = "045e8dbec8c137a52f0f56e196d7abe0";
```
Comment (rozenmd) [inline packages/wrangler/src/docs/index.ts:27]:
```suggestion
		const key = process.env.ALGOLIA_PUBLIC_KEY
```

File: .changeset/six-toes-yawn.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": patch
```
Comment (penalosa) [inline .changeset/six-toes-yawn.md:2]: This should probably be a minor

File: packages/wrangler/.gitguardian.yml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+version: 2
+
+secret:
+  # Ignore security incidents with the SHA256 of the occurrence obtained at output or the secret itself
+  ignored-matches:
+    - name:
+      match: 1bf59277ccfa189234f42f9c82a1194c0ef1d10fbbcf9a7ddb06116c48b77cd0
+    - name:
+      match: 4402bac75111bcbc2ee399dd2cec0411ef8dff5bbcf2cef3f58bc4ddbc5882b4
```
Comment (penalosa) [inline packages/wrangler/.gitguardian.yml:9]: If this didn't work, could you remove this file?

File: packages/wrangler/src/docs/index.ts [hunk 7]
```diff
@@ -1,89 +1,53 @@
...
-	if (args.command === "api") {
-		//if api, take them to the API docs
-		urlToOpen = "https://developers.cloudflare.com/workers/wrangler/api/";
-	} else if (args.command && isValidParam(args.command)) {
-		//otherwise, they get the wrangler commands page
-		urlToOpen += `#${argToUrlHash[args.command]}`;
+	let urlToOpen = "";
+
+	if (args.command) {
+		const id = process.env.ALGOLIA_APP_ID || "";
```
Comment (penalosa) [inline packages/wrangler/src/docs/index.ts:25]:
```suggestion
		const id = process.env.ALGOLIA_APP_ID ?? "";
```

I feel like an error should be thrown if `ALGOLIA_APP_ID` or `ALGOLIA_PUBLIC_KEY` aren't present—there's no change of recovery in that case. Also—does Wrangler's build step bundle the env into `process.env`?

File: packages/wrangler/src/docs/index.ts [hunk 8]
```diff
@@ -1,89 +1,53 @@
...
+		const index = "developers-cloudflare2";
```
Comment (penalosa) [inline packages/wrangler/src/docs/index.ts:26]: Why the 2? Is that what the docs uses?

File: packages/wrangler/src/docs/index.ts [hunk 9]
```diff
@@ -1,89 +1,53 @@
...
+		const key = process.env.ALGOLIA_PUBLIC_KEY || "";
```
Comment (penalosa) [inline packages/wrangler/src/docs/index.ts:27]: As above?

File: packages/wrangler/src/docs/index.ts [hunk 10]
```diff
@@ -1,89 +1,53 @@
...
+		const searchResp = await fetch(
+			`https://${id}-dsn.algolia.net/1/indexes/${index}/query`,
+			{
+				method: "POST",
+				body: JSON.stringify({
+					params: `query=${args.command}&hitsPerPage=1&getRankingInfo=0`,
+				}),
```
Comment (penalosa) [inline packages/wrangler/src/docs/index.ts:34]: The Algolia API takes a JSON encoded object with a URL encoded `param` object???

File: packages/wrangler/src/docs/index.ts [hunk 11]
```diff
@@ -1,89 +1,53 @@
...
+	if (args.command) {
+		const id = process.env.ALGOLIA_APP_ID || "";
+		const index = "developers-cloudflare2";
+		const key = process.env.ALGOLIA_PUBLIC_KEY || "";
+		const searchResp = await fetch(
+			`https://${id}-dsn.algolia.net/1/indexes/${index}/query`,
+			{
+				method: "POST",
+				body: JSON.stringify({
+					params: `query=${args.command}&hitsPerPage=1&getRankingInfo=0`,
```
Comment (penalosa) [inline packages/wrangler/src/docs/index.ts:33]: `args.command` should be URL encoded—probably `new URLSearchParams().toString()` would be best here?

File: packages/wrangler/src/docs/index.ts [hunk 12]
```diff
@@ -1,89 +1,53 @@
...
+				}),
+				headers: {
+					"X-Algolia-API-Key": key,
+					"X-Algolia-Application-Id": id,
+				},
+			}
+		);
+		const searchData = (await searchResp.json()) as { hits: { url: string }[] };
```
Comment (penalosa) [inline packages/wrangler/src/docs/index.ts:41]: Could you add a check for `searchResp.ok`?

File: packages/wrangler/src/docs/index.ts [hunk 13]
```diff
@@ -1,89 +1,53 @@
...
+		logger.debug("searchData: ", searchData);
+		if (searchData.hits[0]) {
+			urlToOpen = searchData.hits[0].url;
+		} else {
+			logger.error(
+				`Could not find docs for: ${args.command}. Please try again with another search term.`
```
Comment (penalosa) [inline packages/wrangler/src/docs/index.ts:47]: Is it worth including a link to the developer docs Wrangler page here as a fallback?

File: packages/wrangler/src/docs/index.ts (hunk 8 above)
Comment (rozenmd) [reply packages/wrangler/src/docs/index.ts:26]: Yep

File: packages/wrangler/src/docs/index.ts (hunk 10 above)
Comment (rozenmd) [reply packages/wrangler/src/docs/index.ts:34]: You love to see it: 
image

File: packages/wrangler/src/index.ts [hunk 14]
```diff
@@ -271,7 +271,7 @@ export function createCLIParser(argv: string[]) {
 
 	// docs
 	wrangler.command(
-		"docs [command]",
+		"docs [command..]",
```
Comment (rozenmd) [inline packages/wrangler/src/index.ts:274]: @GregBrimble I figured it out - this lets you have an optional command that takes an array

so `wrangler docs pages astro` works now

Comment (rozenmd) [reply packages/wrangler/src/index.ts:274]: or maybe it was @WalshyDev, forgot which one I showed this to lol

**Comment by rozenmd** [comment]
Just awaiting the keys to be added to the Github secrets before this can be merged

File: packages/wrangler/src/index.ts (hunk 14 above)
Comment (WalshyDev) [reply packages/wrangler/src/index.ts:274]: Oh hell yes. Ty Max!

File: packages/wrangler/src/docs/helpers.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
+import { fetch } from "undici";
+import { logger } from "../logger";
+
+export async function runSearch(searchTerm: string) {
+	const id = process.env.ALGOLIA_APP_ID || "";
```
Comment (rozenmd) [inline packages/wrangler/src/docs/helpers.ts:5]: TODO: chuck an assert here

**Review by penalosa — APPROVED** [review]

**Pull request merged**